### PR TITLE
Implement `schemaValidation` mode handling in `dataconnect:sql:migrate` and `deploy`

### PR DESCRIPTION
**Pull request by rosalyntan**
### Description

Updates `firebase dataconnect:sql:migrate` and `firebase deploy` per go/fdc-compat-mode-cli-api.

### Scenarios Tested

**`schemaValidation` unset, both required and optional changes:**

- [x] `firebase deploy`
- [x] `firebase deploy --force`
- [x] `firebase deploy --non-interactive`
- [x] `firebase deploy --non-interactive --force`
- [x] `firebase dataconnect:sql:migrate`
- [x] `firebase dataconnect:sql:migrate --force`
- [x] `firebase dataconnect:sql:migrate --non-interactive`
- [x] `firebase dataconnect:sql:migrate --non-interactive --force`

**`schemaValidation` unset, only required changes:**

- [x] `firebase deploy`
- [x] `firebase deploy --force`
- [x] `firebase deploy --non-interactive`
- [x] `firebase deploy --non-interactive --force`
- [x] `firebase dataconnect:sql:migrate`
- [x] `firebase dataconnect:sql:migrate --force`
- [x] `firebase dataconnect:sql:migrate --non-interactive`
- [x] `firebase dataconnect:sql:migrate --non-interactive --force`

**`schemaValidation` unset, only optional changes:**

- [x] `firebase deploy`
- [x] `firebase deploy --force`
- [x] `firebase deploy --non-interactive`
- [x] `firebase deploy --non-interactive --force`
- [x] `firebase dataconnect:sql:migrate`
- [x] `firebase dataconnect:sql:migrate --force`
- [x] `firebase dataconnect:sql:migrate --non-interactive`
- [x] `firebase dataconnect:sql:migrate --non-interactive --force`

**`schemaValidation=COMPATIBLE`, both required and optional changes:**

- [x] `firebase deploy`
- [x] `firebase deploy --force`
- [x] `firebase deploy --non-interactive`
- [x] `firebase deploy --non-interactive --force`
- [x] `firebase dataconnect:sql:migrate`
- [x] `firebase dataconnect:sql:migrate --force`
- [x] `firebase dataconnect:sql:migrate --non-interactive`
- [x] `firebase dataconnect:sql:migrate --non-interactive --force`

**`schemaValidation=COMPATIBLE`, only optional changes:**

- [x] `firebase deploy`
- [x] `firebase deploy --force`
- [x] `firebase deploy --non-interactive`
- [x] `firebase deploy --non-interactive --force`
- [x] `firebase dataconnect:sql:migrate`
- [x] `firebase dataconnect:sql:migrate --force`
- [x] `firebase dataconnect:sql:migrate --non-interactive`
- [x] `firebase dataconnect:sql:migrate --non-interactive --force`

**`schemaValidation=STRICT`, both required and optional changes:**

- [x] `firebase deploy`
- [x] `firebase deploy --force`
- [x] `firebase deploy --non-interactive`
- [x] `firebase deploy --non-interactive --force`
- [x] `firebase dataconnect:sql:migrate`
- [x] `firebase dataconnect:sql:migrate --force`
- [x] `firebase dataconnect:sql:migrate --non-interactive`
- [x] `firebase dataconnect:sql:migrate --non-interactive --force`

### Sample Commands

`firebase deploy`, with `schemaValidation` unset, both required and optional changes, and default options selected:
Screenshot 2024-09-13 at 6 42 25 PM

### CODE REVIEW

File: src/commands/dataconnect-sql-migrate.ts [hunk 1]
```diff
@@ -34,6 +34,7 @@ export const command = new Command("dataconnect:sql:migrate [serviceId]")
       options,
       schema: serviceInfo.schema,
       validateOnly: true,
+      schemaValidation: serviceInfo.dataConnectYaml.schema.datasource.postgresql?.schemaValidation,
```
Comment (joehan) [inline src/commands/dataconnect-sql-migrate.ts:37]: Quick sanity check - does this get broken by our move from datasource -> datasources? @fredzqm might know off the top of his head.

File: src/dataconnect/schemaMigration.ts [hunk 2]
```diff
@@ -131,8 +128,13 @@ export async function migrateSchema(args: {
     databaseId,
     /* linkIfNotConnected=*/ true,
   );
+  let diffs: Diff[] = [];
 
-  const validationMode = experiments.isEnabled("fdccompatiblemode") ? "COMPATIBLE" : "STRICT";
+  let validationMode: SchemaValidation = "STRICT";
```
Comment (joehan) [inline src/dataconnect/schemaMigration.ts:133]: Nit: Can simplify this whole chunk to:
```
const validationMode: SchemaValidation = experiments.isEnabled('fdccompatiblemode') ? (schemaValidation ?? "COMPATIBLE") : "STRICT";
```

File: src/dataconnect/schemaMigration.ts [hunk 3]
```diff
@@ -184,9 +185,50 @@ export async function migrateSchema(args: {
       // Then, try to upsert schema again. If there still is an error, just throw it now
       await upsertSchema(schema, validateOnly);
     }
-    return diffs;
   }
-  return [];
+
+  if (experiments.isEnabled("fdccompatiblemode")) {
+    // If the validation mode is unset, then we also surface any additional optional STRICT diffs.
+    if (!schemaValidation) {
```
Comment (joehan) [inline src/dataconnect/schemaMigration.ts:192]: Nit: Can combine these top 2 conditonals:
```
if (experiments.isEnabled("fdccompatiblemode") && !schemaValidation)
```

**Review by joehan — APPROVED** [review]

File: src/commands/dataconnect-sql-migrate.ts (hunk 1 above)
Comment (rosalyntan) [reply src/commands/dataconnect-sql-migrate.ts:37]: Chatted offline -- based on go/fdc:multi-data-source it looks like the `datasource` field will still be supported, though we should also support `datasources` in #7589.

File: src/dataconnect/schemaMigration.ts (hunk 2 above)
Comment (rosalyntan) [reply src/dataconnect/schemaMigration.ts]: Done.

File: src/dataconnect/schemaMigration.ts (hunk 3 above)
Comment (rosalyntan) [reply src/dataconnect/schemaMigration.ts]: Done.

**Pull request merged**